microcom.1
'\"
'\" Copyright (C) 2009 Alexander Reichle-Schmehl <tolimar@debian.org>
'\" Copyright (C) 2013 Uwe Kleine-König <uwe@kleine-koenig.org>
.TH "MICROCOM" "1"
.SH "NAME"
microcom \- A minimalistic terminal program
.SH "SYNOPSIS"
.PP
.B microcom
.RB [ -d ]
.RB [ -f ]
.RB [\| \-p
.IR devfile |\| \fB\-t
.IR host : port ]
.RB [\| \-s
.IR speed \|]
.RB [\| \-e
.IR escape-char \|]
.br
.B microcom -c
.IB interface : rx_id : tx_id
.br
.B microcom -h

.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBmicrocom\fR command.
.PP
\fBmicrocom\fR is a is a minimalistic terminal
program for accessing devices (e.g. switches) via a serial connection.
It features connection via RS232 serial interfaces (including setting of
transferrates) as well as in `telnetmode' as specified in
rfc2217 and a (Linux specific) can mode.
.PP
The escape character to enter the program menu is crtl-\e. There are several
commands available, among them
.B quit
(to exit microcom),
.B exit
(to return to normal mode) and
.B speed
(to set terminal speed)

.SH "OPTIONS"
.PP
.B microcom
follows the usual GNU command line syntax, with long options starting with two
dashes (`\-'). A summary of options is included below.
.TP
.B \-d
output debugging info to stdout.
.TP
.B \-f
ignore an already existing lock file.
.TP
.BI \-p\  devfile \fR,\ \fB\-\-port= devfile
use the specified serial port device (default
.BR /dev/ttyS0 ).
.TP
.BI \-s\  speed \fR,\ \fB\-\-speed= speed
use specified baudrate (default \fB115200\fR).
.TP
.BI \-t\  host\fB:\fIport \fR,\ \fB\-\-telnet= host\fB:\fIport
work in telnet (rfc2217) mode.
.TP
.BI \-c\  interface\fB:\fIrx_id\fB:\fItx_id\fR,\ \fI \-\-can= interface\fB:\fIrx_id\fB:\fItx_id
work in CAN mode (default: \fBcan0:200:200\fR)
.TP
.BI \-e\  escape-character \fR,\ \fB\-\-escape-char= char
use specified escape character with Ctrl (default \fB\\\fR).
.TP
.BR -h ", " \-\-help
Show help.

.SH "AUTHOR"
.PP
This manual page was written by Uwe Kleine-K\(:onig based on work initially
done by Alexander Reichle-Schmehl.


ChangeLog
2014-09-22  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * Makefile: release 2014.09.0

2014-09-11  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * telnet.c: make telnet mode address family agnostic  Use getaddrinfo et al
  over gethostbyname. This allows (depending on system configuration) to use
  ipv6 connections. As an added bonus allow specifying ipv6 addresses in the
  usual square bracketed syntax.

2014-02-18  Sascha Hauer <s.hauer@pengutronix.de>

  * mux.c: Only send answerback string if one is set

2014-02-12  Sascha Hauer <s.hauer@pengutronix.de>

  * mux.c: factor out handle_receive_buf function  Instead of handling special
  characters in two places.

  * microcom.c, microcom.h, mux.c: Add ENQ/answerback support  This adds
  support for a configurable answerback string for an enquiry character (ENQ,
  ASCII 0x5). The answerback string can be specified using the --answerback
  option. If this option is not given, no answerback string is sent which is
  the behaviour without this patch.

  * microcom.c: update help text  Document the long options and some missing
  options.

2014-02-11  Sascha Hauer <s.hauer@pengutronix.de>

  * mux.c: remove unnecessary return

2014-02-11  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * ChangeLog, Makefile: release 2014.02.0

2014-02-11  Marc Kleine-Budde <mkl@pengutronix.de>

  * microcom.c: microcom: add more baudrates  This patch adds support for more
  baudrates.  Acked-by: Sascha Hauer <s.hauer@pengutronix.de>

  * commands.c, microcom.c, microcom.h: microcom: fix check if
  baudrate_to_flag() fails  baudrate_to_flag() returns a "speed_t" and -1 in
  case of an error. As speed_t is usually unsinged int an error is never
  detected.  This patch converts baudrate_to_flag() to pass speed_t via
  reference, so that the return value can be used to indicate an error. 
  Acked-by: Sascha Hauer <s.hauer@pengutronix.de>

2014-02-11  Sascha Hauer <s.hauer@pengutronix.de>

  * microcom.c: check mux_loop return value, exit with error  If mux_loop
  returns with an error report it and exit microcom with an error code.

  * microcom.h, mux.c: let mux_loop return an error  Instead of void.

  * mux.c: use -1 as invalid filedescriptor  0 is a valid filedescriptor, so
  don't use it to indicate a logfile not being opened.

2014-02-06  Sascha Hauer <s.hauer@pengutronix.de>

  * microcom.c, microcom.h, mux.c: Add support for listononly option  With the
  listenonly option microcom will only listen on the port, but never send
  anything back. Also the local terminal is not modified. This is useful for
  noninteractively creating logfiles from remote ports.

  * commands.c, microcom.c, microcom.h, mux.c: Add logfile support  logfile
  support was lost during restructuring earlier. Add it again.

2014-01-13  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * ChangeLog, Makefile: release 2014.01.1

  * microcom.c: Fix compiler warnings and segfault  Commit 3337bd893b6b dropped
  the comma at the end of the format string by mistake resulting in several
  compiler warnings:  microcom.c:150:3: warning: format ‘%s’ expects
  argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat]
  microcom.c:150:3: warning: format ‘%d’ expects argument of type
  ‘int’, but argument 4 has type ‘char *’ [-Wformat] microcom.c:150:3:
  warning: format ‘%s’ expects argument of type ‘char *’, but argument
  5 has type ‘int’ [-Wformat] microcom.c:150:3: warning: format ‘%x’
  expects a matching ‘unsigned int’ argument [-Wformat]  and a segfault
  when main_usage is called. So readd that comma.  Fixes: 3337bd893b6b (fix
  help string to include -d and -f)

  * .gitignore: .gitignore: don't ignore .gitignore, anchor microcom entry

  * commands.c, commands_fsl_imx.c, parser.c: Add explicit copyright and
  license to three files  These files were only touched by Sascha and Marc back
  in 2010.  Acked-by: Marc Kleine-Budde <mkl@pengutronix.de> Acked-by: Sascha
  Hauer <s.hauer@pengutronix.de>

2014-01-10  Uwe Kleine-König <uwe@kleine-koenig.org>

  * ChangeLog, Makefile: release 2014.01.0

  * gitlog2changelog.py: gitlog2changelog: adapt to output generated by newer
  git  ... that's why you shouldn't parse output of porcelain tools

2014-01-04  Uwe Kleine-König <uwe@kleine-koenig.org>

  * microcom.c: fix help string to include -d and -f  Also drop wrong comment
  about session logging

2013-12-30  Uwe Kleine-König <uwe@kleine-koenig.org>

  * microcom.1: add a simple manpage  The text bases on work done by Alexander
  Reichle-Schmehl for Debian, licensed under GPL v2+. I converted it from sgml
  to plain roff and fixed several things.  Closes:
  http://bugs.debian.org/698436

2013-06-10  Markus Pargmann <mpa@pengutronix.de>

  * Makefile: Make: Fix default link rule for --as-needed  microcom does not
  compile with LDFLAGS="-Wl,--as-needed" because of a default make rule with
  LDFLAGS before all object files. The linker seems to parse the commandline
  only once, which results in this case in undefined references.  This patch
  defines a custom link rule with LDFLAGS at the end of the command line.

2013-06-10  Michael Olbrich <m.olbrich@pengutronix.de>

  * microcom.c: add signal handler for SIGQUIT  it seems readline() send
  SIGQUIT on ctrl-\

2013-04-29  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * mux.c: Fix a few typos in comments and point out a layer violation

  * microcom.c: fix newline handling  Disable any nl <-> cr translation. This
  makes it possible to drive a debian installer and still directing barebox and
  Linux continues to work.  While at it drop cleaning BRKINT as it doesn't do
  anything with IGNBRK set.

2012-06-14  Marc Kleine-Budde <mkl@pengutronix.de>

  * ChangeLog, Makefile: release 2012.06.0

2012-02-18  Jonas Zetterberg <jozz@jozz.se>

  * serial.c: Remove lockfile on open serial failure  Since the lockfile is
  created before the serial device is opened, if the serial device can not be

microcom.c
// SPDX-License-Identifier: GPL-2.0-only
// SPDX-FileCopyrightText: 1999 Anca and Lucian Jurubita <ljurubita@hotmail.com>.
#define _GNU_SOURCE
#include "microcom.h"

#include <unistd.h>
#include <getopt.h>
#include <sys/socket.h>
#include <string.h>
#include <stdint.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "config.h"

static struct termios sots;     /* old stdout/in termios settings to restore */

struct ios_ops *ios;
int debug = 0;

void init_terminal(void)
{
	struct termios sts;

	memcpy(&sts, &sots, sizeof(sots));      /* to be used upon exit */

	/* Implement what `stty raw` does. */
	sts.c_iflag &= ~(IGNBRK | BRKINT | IGNPAR | PARMRK | INPCK |
			 ISTRIP | INLCR | IGNCR | ICRNL | IXON | IXOFF |
			 IUCLC | IXANY | IMAXBEL);
	sts.c_lflag &= ~(ICANON | ISIG | XCASE);
	sts.c_oflag &= ~OPOST;
	sts.c_cc[VMIN] = 1;
	sts.c_cc[VTIME] = 0;

	/* no local echo: allow the other end to do the echoing */
	sts.c_lflag &= ~ECHO;

	tcsetattr(STDIN_FILENO, TCSANOW, &sts);
}

void restore_terminal(void)
{
	tcsetattr(STDIN_FILENO, TCSANOW, &sots);
}


void microcom_exit(int signal)
{
	printf("exiting\n");

	ios->exit(ios);
	tcsetattr(STDIN_FILENO, TCSANOW, &sots);

	exit(0);
}

/*
 * Main functions
 ********************************************************************
 * static void help_usage(int exitcode, char *error, char *addl)
 *      help with running the program
 *      - exitcode - to be returned when the program is ended
 *      - error - error string to be printed
 *      - addl - another error string to be printed
 * static void cleanup_termios(int signal)
 *      signal handler to restore terminal set befor exit
 * int main(int argc, char *argv[]) -
 *      main program function
 */
void main_usage(int exitcode, char *str, char *dev)
{
	fprintf(stderr, "Usage: microcom [options]\n"
		" [options] include:\n"
		"    -p, --port=<devfile>                 use the specified serial port device (%s);\n"
		"    -s, --speed=<speed>                  use specified baudrate (%d)\n"
		"    -t, --telnet=<host:port>             work in telnet (rfc2217) mode\n"
		"    -c, --can=<interface:rx_id:tx_id>    work in CAN mode\n"
		"                                         default: (%s:%x:%x)\n"
		"    -f, --force                          ignore existing lock file\n"
		"    -d, --debug                          output debugging info\n"
		"    -l, --logfile=<logfile>              log output to <logfile>\n"
		"    -o, --listenonly                     Do not modify local terminal, do not send input\n"
		"                                         from stdin\n"
		"    -a, --answerback=<str>               specify the answerback string sent as response to\n"
		"                                         an ENQ (ASCII 0x05) Character\n"
		"    -e, --escape-char=<chr>              escape charater to use with Ctrl (%c)\n"
		"    -v, --version                        print version string\n"
		"    -h, --help                           This help\n",
		DEFAULT_DEVICE, DEFAULT_BAUDRATE,
		DEFAULT_CAN_INTERFACE, DEFAULT_CAN_ID, DEFAULT_CAN_ID,
		DEFAULT_ESCAPE_CHAR);
	fprintf(stderr, "Exitcode %d - %s %s\n\n", exitcode, str, dev);
	exit(exitcode);
}

int opt_force = 0;
unsigned long current_speed = DEFAULT_BAUDRATE;
int current_flow = FLOW_NONE;
int listenonly = 0;
char escape_char = DEFAULT_ESCAPE_CHAR;

int main(int argc, char *argv[])
{
	struct sigaction sact = { 0 };  /* used to initialize the signal handler */
	int opt, ret;
	char *hostport = NULL;
	int telnet = 0, can = 0;
	char *interfaceid = NULL;
	char *device = DEFAULT_DEVICE;
	char *logfile = NULL;

	struct option long_options[] = {
		{ "help", no_argument, NULL, 'h' },
		{ "port", required_argument, NULL, 'p' },
		{ "speed", required_argument, NULL, 's' },
		{ "telnet", required_argument, NULL, 't' },
		{ "can", required_argument, NULL, 'c' },
		{ "debug", no_argument, NULL, 'd' },
		{ "force", no_argument, NULL, 'f' },
		{ "logfile", required_argument, NULL, 'l' },
		{ "listenonly", no_argument, NULL, 'o' },
		{ "answerback", required_argument, NULL, 'a' },
		{ "version", no_argument, NULL, 'v' },
		{ 0, 0, 0, 0 },
	};

	while ((opt = getopt_long(argc, argv, "hp:s:t:c:dfl:oi:a:e:v", long_options, NULL)) != -1) {
		switch (opt) {
		case '?':
			main_usage(1, "", "");
			break;
		case 'h':
			main_usage(0, "", "");
			break;
		case 'v':
			printf("%s\n", PACKAGE_VERSION);
			exit(EXIT_SUCCESS);
			break;
		case 'p':
			device = optarg;
			break;
		case 's':
			current_speed = strtoul(optarg, NULL, 0);
			break;
		case 't':
			telnet = 1;
			hostport = optarg;
			break;
		case 'c':
			can = 1;
			interfaceid = optarg;
			break;
		case 'f':
			opt_force = 1;
			break;
		case 'd':
			debug = 1;
			break;
		case 'l':
			logfile = optarg;
			break;
		case 'o':
			listenonly = 1;
			break;
		case 'a':
			answerback = (unsigned char *)optarg;
			break;
		case 'e':
			if (strlen(optarg) != 1) {
				fprintf(stderr, "Option -e requires a single character argument.\n");
				exit(EXIT_FAILURE);
			}
			escape_char = *optarg;
			break;
		}
	}

	if (optind < argc)
		main_usage(1, "", "");

	if (answerback) {
		ret = asprintf((char **)&answerback, "%s\n", answerback);
		if (ret < 0)
			exit(1);
	}

	commands_init();
	commands_fsl_imx_init();

	if (telnet && can)
		main_usage(1, "", "");

	if (telnet)
		ios = telnet_init(hostport);
	else if (can) {
#ifdef USE_CAN
		ios = can_init(interfaceid);
#else
		fprintf(stderr, "CAN mode not supported\n");
		exit(EXIT_FAILURE);
#endif
	} else
		ios = serial_init(device);

	if (!ios)
		exit(1);

	if (logfile) {
		ret = logfile_open(logfile);
		if (ret < 0)
			exit(1);
	}

	ret = ios->set_speed(ios, current_speed);
	if (ret)
		goto cleanup_ios;

	current_flow = FLOW_NONE;
	ios->set_flow(ios, current_flow);

	if (!listenonly) {
		printf("Escape character: Ctrl-%c\n", escape_char);
		printf("Type the escape character to get to the prompt.\n");

		/* Now deal with the local terminal side */
		tcgetattr(STDIN_FILENO, &sots);
		init_terminal();

		/* set the signal handler to restore the old
		 * termios handler */
		sact.sa_handler = &microcom_exit;
		sigaction(SIGHUP, &sact, NULL);
		sigaction(SIGINT, &sact, NULL);
		sigaction(SIGPIPE, &sact, NULL);
		sigaction(SIGTERM, &sact, NULL);
		sigaction(SIGQUIT, &sact, NULL);
	}

	/* run the main program loop */
	ret = mux_loop(ios);

	if (!listenonly)
		tcsetattr(STDIN_FILENO, TCSANOW, &sots);

cleanup_ios:
	ios->exit(ios);

	exit(ret ? 1 : 0);
}
